libs/libcommon/src/libcommon/url_signer.py

```diff
@@ -0,0 +1,131 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2024 The HuggingFace Authors.
+
+from abc import ABC, abstractmethod
+from collections.abc import Mapping
+from dataclasses import dataclass
+from typing import Any, Callable, Literal, Optional, Union
+
+from datasets import Audio, Features, Image
+from datasets.features.features import FeatureType, Sequence
+
+from libcommon.dtos import FeatureItem
+
+
+class InvalidFirstRowsError(ValueError):
+    pass
+
+
+VisitPath = list[Union[str, Literal[0]]]
+
+
+@dataclass
+class AssetUrlPath:
+    feature_type: Literal["Audio", "Image"]
+    path: VisitPath
+
+    def enter(self) -> "AssetUrlPath":
+        if len(self.path) == 0:
+            raise ValueError("Cannot enter an empty path")
+        return AssetUrlPath(feature_type=self.feature_type, path=self.path[1:])
+
+
+# invert to_features_list
+def to_features_dict(features: list[FeatureItem]) -> Features:
+    return Features({feature_item["name"]: feature_item["type"] for feature_item in features})
+
+
+def _visit(
+    feature: FeatureType, func: Callable[[FeatureType, VisitPath], Optional[FeatureType]], visit_path: VisitPath = []
+) -> FeatureType:
+    """Visit a (possibly nested) feature.
+
+    Args:
+        feature (FeatureType): the feature type to be checked
+    Returns:
+        visited feature (FeatureType)
+    """
+    if isinstance(feature, dict):
```

Comment:
There are two possibilities in Sequence:

```
Sequence(Image())  # e.g. [image0, image1, ...]
```

or 

```
Sequence({"images": Image()})  # e.g. {"images": [image0, image1, ...]}
```

you can take this into account by adding these two lines:

```suggestion
    if isinstance(feature, Sequence) and isinstance(feature.feature, dict):
            feature = {k: [f] for k, f in feature.items()}
    if isinstance(feature, dict):
```

the original `_visit` in `datasets` doesn't need this since it doesn't keep track of the path

Reply:
I sent https://github.com/huggingface/datasets-server/pull/2315/commits/2ca0df11e5fd89333ab49d38ce897aff7c175b59, but I think I didn't get it right, because the tests pass with or without these two lines at the start of `_visit`.

Could you have a look @lhoestq?

Reply:
added a comment in your commit

Reply:
thanks! The issue was: that I hadn't enabled my test